### ISSUE TITLE
QUIC: Add API to allow detecting when peer aborts receive

### ISSUE THREAD

**Issue by geoffkizer** [dotnet/runtime#58229]
Currently, the only way to detect that a peer has aborted receiving is to try to send and see if it fails.

If a user is periodically sending data on the QuicStream, for example from a backend SQL query or other source, this means they will not discover the abort until they finish the query and try to write to the QuicStream. If this is expensive and/or long-running, that could be problematic. It would be better if they could discover the peer abort promptly and cancel any in-progress work.

This happens specifically in ASP.NET with a "streamed response", where the request handling logic is writing to the response stream periodically. Currently, Kestrel has no way to be notified that the HTTP client has aborted the response.

To address this, we should add an API like so:

```csharp
    public Task WaitForWriteCompletionAsync(CancellationToken cancellationToken);
```
 
If the peer aborts writes (or closes the connection, or there’s a connection failure), this would throw the appropriate error (QuicStreamAbortedException or QuicConnectionAbortedException, etc).
If the write stream completes successfully, this would return without exception.
 
The logic to implement this should be similar to the old ShutdownWriteCompleted method that got removed here: https://github.com/dotnet/runtime/pull/55981. This code would detect when the write stream completed successfully.

However, note that the old ShutdownWriteCompleted code does not seem to have handled the peer receive abort case. We will need to add code in HandleEventPeerRecvAborted to do this.

**Comment by karelz** [comment]
Reopening to track backport to 6.0 branch

**Comment by karelz** [comment]
Fixed in 6.0 RC2 in PR #58415.
Fixed in 7.0 in PR #58236.
